2022Swerve/src/main/java/frc/robot/subsystems/AutonSubsytem.java

```diff
@@ -0,0 +1,193 @@
+    private StringLogEntry commandLog = new StringLogEntry(datalog, "/auton/command"); //Logs x translation state output
+
+    private SwerveSubsystem swerveSubsystem;
+
+    private PIDController xController;
+    private PIDController yController;
+    private PIDController rotationController;
+
+    private PathConstraints pathConstraints;
+
+    private AutonModes autonMode;
+
+    public AutonSubsytem(SwerveSubsystem swerveSubsystem){
+        this.swerveSubsystem = swerveSubsystem;
+
+        //Add auton modes to chooser
+        for(AutonModes mode : AutonModes.values()){
+            modeChooser.addOption(mode.toString(), mode);
+        }
+        modeChooser.setDefaultOption(kDefaultAutonMode.toString(), kDefaultAutonMode);
+        configTab.add("Auton mode", modeChooser);
+
+        //Define PID controllers for tracking trajectory
+        xController = new PIDController(AutonConstants.kXTranslationP, 0, 0);
+        yController = new PIDController(AutonConstants.kYTranslationP, 0, 0);
+        rotationController = new PIDController(AutonConstants.kRotationP, 0, 0);
+        rotationController.enableContinuousInput(-Math.PI, Math.PI);
+
+        pathConstraints = new PathConstraints(AutonConstants.kMaxSpeed, AutonConstants.kMaxAcceleration);
+    }
+
+    private PathPlannerTrajectory getTrajectory(String name) throws NullPointerException{
+        return PathPlanner.loadPath(name, pathConstraints);
+    }
+
+    private PathPoint constructPoint(double x, double y, double rotation, double heading){
+        return new PathPoint(new Translation2d(x, y), Rotation2d.fromDegrees(heading), Rotation2d.fromDegrees(rotation));
+    }
+
+    private PathPlannerTrajectory generateTrajectory(PathPoint firstPoint, PathPoint secondPoint, PathPoint... points){
+        return PathPlanner.generatePath(
+            pathConstraints,
+            firstPoint,
+            secondPoint,
+            points
+        );
+    }
+
+    private PPSwerveControllerCommand followTrajectory(String name) throws NullPointerException{
+        //For use with trajectories generated from pathplanner
+        trajectoryLog.append("Following trajectory " + name);
+        PathPlannerTrajectory trajectory = getTrajectory(name);
+        if(Objects.isNull(trajectory)){throw new NullPointerException();}
+        return new PPSwerveControllerCommand(
+            trajectory,
+            swerveSubsystem::getPose, 
+            DriveConstants.kDriveKinematics, 
+            xController, 
+            yController, 
+            rotationController, 
+            swerveSubsystem::setModuleStates, 
+            swerveSubsystem
+        );
+    }
+
+    private PPSwerveControllerCommand followTrajectory(PathPlannerTrajectory trajectory){
+        //For use with trajectories generated from a list of poses
+        trajectoryLog.append("Following generated trajectory");
+        return new PPSwerveControllerCommand(
```

Comment:
WOuld like to see this log immediately before the trajectry command is run, and immediately after it is finished using andThen, and beforeStarting  or two andThen's